yarn-project/circuit-types/src/versioning.ts

```diff
@@ -0,0 +1,152 @@
+import { type EthAddress } from '@aztec/foundation/eth-address';
+import { type Fr } from '@aztec/foundation/fields';
+import { jsonStringify } from '@aztec/foundation/json-rpc';
+
+import type Koa from 'koa';
+
+import { type ChainConfig } from './config.js';
+
+// REFACTOR: This file is not a circuit-type, but at the moment we don't have any other
+// package common to all components that we can use for this shared code.
+
+/** Fields that identify a version of the Aztec protocol. Any mismatch between these fields should signal an incompatibility between nodes. */
+export type ComponentsVersions = {
+  l1ChainId: number;
+  // TODO: Consider using governance address instead to support migrations.
+  // Note that we are using the rollup address as identifier in multiple places
+  // such as the keystore, we need to change it so we can handle updates.
+  l1RollupAddress: EthAddress;
+  l2ChainVersion: number;
+  l2ProtocolContractsTreeRoot: string;
+  l2CircuitsVkTreeRoot: string;
+};
+
+/** Returns components versions from chain config. */
+export function getComponentsVersionsFromConfig(
+  config: ChainConfig,
+  l2ProtocolContractsTreeRoot: string | Fr,
+  l2CircuitsVkTreeRoot: string | Fr,
+): ComponentsVersions {
+  return {
+    l1ChainId: config.l1ChainId,
+    l1RollupAddress: config.l1Contracts?.rollupAddress, // This should not be undefined, but sometimes the config lies to us and it is...
+    l2ChainVersion: config.version,
+    l2ProtocolContractsTreeRoot: l2ProtocolContractsTreeRoot.toString(),
+    l2CircuitsVkTreeRoot: l2CircuitsVkTreeRoot.toString(),
+  };
+}
+
+/** Returns a compressed string representation of the version (around 32 chars). Used in p2p ENRs. */
+export function compressComponentVersions(versions: ComponentsVersions): string {
+  if (
+    versions.l1RollupAddress === undefined ||
+    versions.l2ProtocolContractsTreeRoot === undefined ||
+    versions.l2CircuitsVkTreeRoot === undefined
+  ) {
+    throw new Error(`Component versions are not set: ${jsonStringify(versions)}`);
+  }
+  return [
+    '00',
+    versions.l1ChainId,
+    versions.l1RollupAddress.toString().slice(2, 10),
+    versions.l2ChainVersion,
+    versions.l2ProtocolContractsTreeRoot.toString().slice(2, 10),
+    versions.l2CircuitsVkTreeRoot.toString().slice(2, 10),
+  ].join('-');
+}
+
+export class ComponentsVersionsError extends Error {
+  constructor(key: string, expected: string, value: string) {
+    super(`Expected component version ${key} to be ${expected} but received ${value}`);
+    this.name = 'ComponentsVersionsError';
+  }
+}
+
+/** Checks if the compressed string matches against the expected versions. Throws on mismatch. */
+export function checkCompressedComponentVersion(compressed: string, expected: ComponentsVersions) {
```

Comment:
Rather than checking each, do we want the compresedComponentVersion to just be a fast hash of all of them stuck together ( see https://github.com/AztecProtocol/aztec-packages/blob/756b94101345d231b86233921398555512855273/yarn-project/p2p/src/services/encoding.ts#L11). This will reduce the size of the enr that is being sent around for the discovery protocol, so will have real bandwidth implications